### DISCUSSION TITLE
Custom Heterogenous data from csv

### DISCUSSION THREAD

**Discussion by aaronfderybel** [pyg-team/pytorch_geometric#4655] [Q&A]
Hello,
I'm currently working on a computer networking problem to detect suspicious activity (hackers f.e.)
My dataset in csv format contains a connection for each row between two devices (see figure below)
Each row contains following information
* TimeStamp
* Source and destination IP adresss of devices
* Features related to connections
* Label related to connection (some attack or benign behaviour)
![image](https://user-images.githubusercontent.com/43145159/168591886-4e16b188-f1b0-4316-9155-451937694aa4.png)

What I want to achieve is for each row create the following part of a graph (f.e. for row 1)
* node of type *IP* for source and destination device
* node of type *connection* 
* edge between source IP and connection node
* edge between connection node and destination IP
![image](https://user-images.githubusercontent.com/43145159/168593914-c22df51b-9911-4764-9ef4-ec69e684a89f.png)

I was looking into the example here: https://pytorch-geometric.readthedocs.io/en/latest/notes/load_csv.html
I'm not sure how I would create my edge indices in this case.

```python
data = HeteroData()
data['connection'].x = ... #[num_rows, connection_features], this part is easy
#there are only connections between ip nodes and connection nodes
data['ip','connects', 'connection'].edge_index = torch.Tensor([ [?, ?, ...], [?, ?, ...]])
```
Does anyone have experience with creating this type of graphs? Point my into the right direction?

**Comment by rusty1s** [comment]
I am not sure if that is the best way to model that. IMO, both src IP and dst IP denote a node in a homogeneous graph, and there exists an edge between a link in case there exists a connection between both IPs. Duration, Time first seen, etc are then just edge features in your graph.

**Reply by aaronfderybel** [reply]
Hello @rusty1s ,

This heterogenous graph seems a bit far fetched maybe, i'm basing the construction on this paper: https://arxiv.org/pdf/2107.14756v1.pdf

A connection or flow between two devices in a network is defined by a unique combination of src, dst IP and also src and dst ports.
So you can have multiple connections between devices A and B going on at the same time for different ports. Different ports usually are used for different services (uploading files, http requests, ...)
A portscan is a technique where you check out all the ports of a device (this is used by hackers and network engineers). 
In the graph it looks like this:
![image](https://user-images.githubusercontent.com/43145159/169005866-a3edf148-8a17-4dfe-b045-db2c55ade972.png)

The hacker or engineer (a) is scanning all the ports on a device (v). If you would use homogenous graph you can only have up to two edges between the two IP nodes. I don't think you would be able to include all the connection information this way.

I implemented a class in python to convert the csv file into graphs, each containing 200 connection nodes.
I was following this tutorial on Heterogenous graph training.
https://pytorch-geometric.readthedocs.io/en/latest/notes/heterogeneous.html#deploy-existing-heterogenous-operators
During lazy initialization of the model I receive the following error
```python
model = HGT(data_graph= data[0], hidden_channels=64, out_channels=5,
            num_heads=2, num_layers=1)

with torch.no_grad():  # Initialize lazy modules.
     out = model(data[0].x_dict, data[0].edge_index_dict)
```
![image](https://user-images.githubusercontent.com/43145159/169822500-f33ef6b9-06e1-411a-a698-ca8d5db87507.png)

How should I fix this error?
I've included a notebook+dataset with all the steps for reproducing the error.
[reproduce_error.zip](https://github.com/pyg-team/pytorch_geometric/files/8754678/reproduce_error.zip)

**Reply by rusty1s** [reply]
It looks like your `HeteroData` objects are wrongly encoded. You need to ensure that the following passes:
```python
#sizes seem to be correct, 200 connection nodes for two directions for each graph in the dataset
#ip nodes can vary
data[0]

for d in data:
    for edge_type in d.edge_types:
        src, _, dst = edge_type
        assert d[edge_type].edge_index[0].max() < d[src].num_nodes
        assert d[edge_type].edge_index[1].max() < d[dst].num_nodes 
```

**Reply by aaronfderybel** [reply]
Hi @rusty1s thank you for the feedback and time,
I fixed my code so this assertion passes and I can train my model sucesfully.
You may close this question as answered :)
